### DISCUSSION TITLE
Set Custom Filter in Product Module

### DISCUSSION THREAD

**Discussion by Cluster-Taek** [medusajs/medusa#10614] [Q&A]
Thank you for always providing excellent libraries and documentation.

Following the documentation, we linked brands to products separately from categories, and all features are working flawlessly on the admin page.

However, we are encountering an issue when attempting to apply a brand filter on a product list component, such as paginated-products.tsx, on the user page.

On the frontend, we have confirmed that the SDK module is generally used to retrieve products. We are wondering how to apply the brand filter in this context.
When we try to incorporate the brand filter into the SDK, a Zod error is returned like this screenshot.

<img width="897" alt="스크린샷 2024-12-16 18 23 58" src="https://github.com/user-attachments/assets/b471965f-5c59-4a7b-981b-ca0ab9a980dc" />

Is there a way to directly apply the brand filter using the SDK?
Or would it be necessary to create a new route for this purpose?

We would be sincerely grateful for your assistance.

**Comment by martinerko** [comment]
@shahednasser I have seen this question raised at Discord as well and it's not clear from documentation how to do this.

The existing documentation only mentions how to pass additional linked data to product
https://docs.medusajs.com/learn/customization/extend-features/extend-create-product#2-configure-additional-data-validation
but completely misses on how to filter products by linked data, e.g. finding all products by brand_id.
`GET /admin/products?brand_id=brand_123`

I have spent some time diving into medusa core and could not find answer either.
Are you kindly able to give us an advice? Happy to help to raise a PR and update with docs once we got the correct answer.

Thanks!

**Reply by shahednasser** [reply]
Hi @martinerko , filtering by brand ID using an existing route is not possible. You have to create a custom route that retrieves the products of a brand.

**Reply by martinerko** [reply]
@shahednasser Thanks for prompt answer.
Now regarding the actual implementation of the API route (can’t test this now), would store’s `sdk` or `graph.query` executed on `product` entity allow me to achieve this?
Assuming I need to filter by brand and category (and possibly further by product price range) at the same time.

**Reply by shahednasser** [reply]
You can use `query.graph` to get the brand of a product (and vice-versa) as explained here: https://docs.medusajs.com/learn/customization/extend-features/query-linked-records#approach-2-use-query-to-retrieve-linked-records

As for the `sdk` I'm not sure I understand your question. If you mean whether you can send a request to the custom API route, you can as explained [here](https://docs.medusajs.com/resources/js-sdk#send-requests-to-custom-routes). You can also see it in [this brand example](https://docs.medusajs.com/learn/customization/customize-admin/route#retrieve-brands-from-api-route).

**Reply by martinerko** [reply]
Got it, thanks for explanation.

**Reply by NicolasGorga** [reply]
@shahednasser hey, i’ve been digging through discussions and issues for this!

is there any plan to allow to extend Params core validators for core routes. For example, i see a lot of middlewares are applied to /store/porducts and logic inside the route.

if i want to be able to filter products by brand, and want it to act exactly like the core, then i see it as a bad thing as i would have to recreate all the core logic inside my custom route. 

I think an example implementation of this applied to Brand example in the docs, will make it clear as to why not having this ability to extend the validator, as opposed to POST core routes (where you can) is quite negative.

**Comment by Vineet1877** [comment]
@shahednasser 

Brand and product are different modules, the link concept establishes the relation. now if we want to achieve we have to send two requests to the database first fetch the brand and then fetch the product of that brand, but if I like it in a single query request then it is not possible.

following is the code:

export const GET = async (
req: MedusaRequest,
res: MedusaResponse
) => {
const query = req.scope.resolve("query")
const { data: product } = await query.graph({
entity: "product",
fields: ["*"],
filters:{
brand: {
"name": "XYZ"
}
}
})
res.json({ product})
}

it shows an error of Entity 'Product' does not have property 'brand'.

**Reply by shahednasser** [reply]
Hi @Vineet1877 

As mentioned in the [Query documentation](https://docs.medusajs.com/learn/fundamentals/module-links/query#apply-filters-and-pagination-on-linked-records) you can't filter by a linked record. Refer to the documentation for the alternative approach: https://docs.medusajs.com/learn/fundamentals/module-links/query#apply-filters-and-pagination-on-linked-records

**Comment by Vineet1877** [comment]
In Medusa, the order and product modules are typically connected via a linked table without a database-level foreign key constraint. This approach has two major downsides:

No Data Integrity at the Database Level – Since the relationship isn't enforced at the database level, data consistency is not guaranteed, which can lead to orphaned records.

Inefficient Querying—When filtering data that includes fields from both order and product tables, two separate queries must be made—one for the order table and another for the product table—resulting in additional database requests and slower performance.

Proposed Solution:
I want to modify the Medusa core to establish a direct foreign-key constraint between the order and product tables.

Proposed Change: (Direct Foreign Key in Order Table)
Instead of using a separate linking table, modify the order table to directly reference the product table:

✅ Ensures Data Integrity – Prevents orphan records by enforcing the relationship at the database level.
✅ Improves Query Performance—A single SQL query can now retrieve data from both orders and products without requiring multiple requests.

Looking for Community Feedback
1. I’d love to hear thoughts from the Medusa community on:
2. Possible compatibility concerns with Medusa’s existing architecture.
3. Best practices for implementing this change in a way that aligns with Medusa’s design principles.
4. Any potential drawbacks or alternative solutions?

Would appreciate any insights! 🚀

**Comment by akordzik** [comment]
I have a similar problem that is missing information in the documentation. The section about [querying Linked Records](https://docs.medusajs.com/learn/fundamentals/module-links/query#apply-filters-and-pagination-on-linked-records) only mentions that "You can then apply any filters or pagination configurations." but doesn't demonstrate how to do this exactly. In my case I'm querying a Linked Record of Product and custom entity and I need to apply additional filtering on Product. I've tried many approaches but nothing seems to work. The full query I'm trying to run looks as follows:

```
    const { data, metadata } = await query.graph({
      entity: productTenant.entryPoint,
      fields,
      pagination,
      filters: {
        tenant_id: tenantId,
        product: {
          status: 'published',
        },
      },
    })
```

The error I'm getting:
> Trying to query by not existing property LinkModel.product

And yet the data returned has a structure of:

```
[
  {
    product_id: '...',
    product: {
      ...
      status: 'published'
    }
  }
]
```

**Reply by shahednasser** [reply]
Thank you for pointing this out @akordzik , we need to provide some clarity around this section.

You can pass filters related to the fields in the module link, so that includes the IDs of the linked data models or [any custom columns](https://docs.medusajs.com/learn/fundamentals/module-links/custom-columns). However, as mentioned earlier in the [Query documentation](https://docs.medusajs.com/learn/fundamentals/module-links/query#apply-filters), you can't filter linked models. So you can't apply a filter on the product's status. You'd have to instead retrieve products, apply the filter on their status, and retrieve their tenant as explained [here](https://docs.medusajs.com/learn/fundamentals/module-links/query#retrieve-linked-records).

**Reply by akordzik** [reply]
Hi @shahednasser, thank for your input here.

Yes, I've already tried that approach, but it comes with another problem, unfortunately. Since filtering by linked models does not seem to work, it means I would be able only to select products with tenant, without filtering it. So say, if I apply pagination and query for 10 products, but only 8 of them belong to given tenant, then only 8 products is returned instead of expected 10.

The only feasible workaround that I've found so far is to:
1. Query module link, just like I do in my example
2. Return all product IDs (belonging to given tenant)
3. Make another query with product filters and with additional filter of product IDs (returned by previous query)

It seems to work fine, but performance is my concern here.

**Reply by shahednasser** [reply]
Yes unfortunately that's the only approach available at the moment.

**Comment by bqst** [comment]
Hi everyone,
Following this discussion about filtering products by linked entities, I'd like to get community feedback on how you would implement the following use case:
I have an "Availability" module with the following model:

```typescript
import { model } from "@medusajs/framework/utils"

export enum AvailabilityStatus {
  INCOMING = "incoming",
  IN_STOCK = "in_stock",
  OUT_OF_STOCK = "out_of_stock",
}

const Availability = model.define("availability", {
  id: model.id().primaryKey(),
  product_id: model.text().unique(), // Link to product
  status: model.enum(AvailabilityStatus).default(AvailabilityStatus.INCOMING),
  notes: model.text().nullable(),
})

export default Availability
```

And the following link definition:

```typescript
import { defineLink } from "@medusajs/framework/utils"
import AvailabilityModule from "../modules/availability"
import ProductModule from "@medusajs/medusa/product"

export default defineLink(ProductModule.linkable.product, {
  linkable: AvailabilityModule.linkable.availability,
  deleteCascade: true,
})
```

I want to create a /store/products/available route that returns only products that are either "in_stock" or "incoming".

The following query doesn't return the products as expected:

```typescript
const query = req.scope.resolve(ContainerRegistrationKeys.QUERY)

const { data: availabilities } = await query.graph({
  entity: "availability",
  fields: ["*", "product.*"],
  filters: {
    status: {
      $in: [AvailabilityStatus.IN_STOCK, AvailabilityStatus.INCOMING],
    },
  },
})

```

And as mentioned in this thread, I can't directly filter products by their availability status.

How would you implement this without having to query the database twice? I was previously doing two separate queries, but this creates noticeable latency.

Thanks!
